### PR TITLE
GH-36767: [C++][CI] Fix test failure on i386

### LINKED ISSUE #36767: [C++][CI] DecimalToReal test failure on 32-bit x86 Debian
### Describe the bug, including details regarding any error messages, version, and platform.

See nightly build here:
https://github.com/ursacomputing/crossbow/actions/runs/5593823296/jobs/10227952675#step:6:3255

This can be reproducer locally using:
```
DEBIAN=11 ARCH=i386 archery docker run debian-cpp
```

### Component(s)

C++

### PR DESCRIPTION
**Pull request by pitrou**
### Rationale for this change

Our nightly build on i386 Debian (a 32-bit x86 build) fails with an error in decimal-to-read tests:
https://github.com/ursacomputing/crossbow/actions/runs/5593823296/jobs/10227952675#step:6:3255

### What changes are included in this PR?

Improve error messages by displaying the actual and expected values.

A side effect of this error message improvement is to... fix the test failure, as storing the computation result in a local variable seems to change the computed absolute difference.

This is probably due to x86 FPU rounding shenanigans, as explained here: https://stackoverflow.com/questions/37626687/c-fundamentals-double-variable-not-equal-to-double-expression

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.

* Closes: #36767

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

File: cpp/src/arrow/util/decimal_test.cc [hunk 1]
```diff
@@ -1046,42 +1046,51 @@ using ToDoubleTestParam = ToRealTestParam<double>;
 template <typename Decimal, typename Real>
 void CheckDecimalToReal(const std::string& decimal_value, int32_t scale, Real expected) {
   Decimal dec(decimal_value);
-  ASSERT_EQ(dec.template ToReal<Real>(scale), expected)
-      << "Decimal value: " << decimal_value << " Scale: " << scale;
+  Real actual = dec.template ToReal<Real>(scale);
```
Comment (js8544) [inline cpp/src/arrow/util/decimal_test.cc:1049]: I'm curious: is it possible to set the precision for gtest outputs like `std::setprecision`. The default precision won't be very helpful for small diffs.

File: cpp/src/arrow/util/decimal_test.cc [hunk 2]
```diff
@@ -1046,42 +1046,51 @@ using ToDoubleTestParam = ToRealTestParam<double>;
...
 }
 
 template <typename Decimal, typename Real>
 void CheckDecimalToRealWithinEpsilon(const std::string& decimal_value, int32_t scale,
                                      Real epsilon, Real expected) {
   Decimal dec(decimal_value);
-  ASSERT_TRUE(std::abs(dec.template ToReal<Real>(scale) - expected) <= epsilon)
-      << "Decimal value: " << decimal_value << " Scale: " << scale;
+  Real actual = dec.template ToReal<Real>(scale);
+  ASSERT_TRUE(std::abs(actual - expected) <= epsilon)
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/decimal_test.cc:1071]: Ah, right. I wonder if it reintroduces the failure :-D

File: cpp/src/arrow/util/decimal_test.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/util/decimal_test.cc:1049]: I wonder if that's necessary. Introducing a deliberate error seems to display the values with sufficient precision:
```
[ RUN      ] TestDecimalToRealFloat/0.Precision
/arrow/cpp/src/arrow/util/decimal_test.cc:1071: Failure
Value of: std::abs(actual - expected) <= epsilon
  Actual: false
Expected: true
Decimal value: 1.987748987892758765582589910934859345, scale: 36, expected: 1.9877489805221558, actual: 1.9877490997314453
[  FAILED  ] TestDecimalToRealFloat/0.Precision, where TypeParam = arrow::Decimal128 (0 ms)
```

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

File: cpp/src/arrow/util/decimal_test.cc (hunk 1 above)
Comment (js8544) [reply cpp/src/arrow/util/decimal_test.cc:1049]: I see. It looks ok to me!

**Review by js8544 — APPROVED** [review]

**Pull request merged**